### ISSUE TITLE
A data race is found in the CI

### ISSUE THREAD

**Issue by tiancaiamao** [pingcap/tidb#11693]
## Bug Report

Please answer these questions before submitting your issue. Thanks!

1. What did you do?
If possible, provide a recipe for reproducing the error.

Found here https://github.com/pingcap/tidb/pull/11675

```
[2019-08-09T05:47:59.044Z] ==================
[2019-08-09T05:47:59.044Z] WARNING: DATA RACE
[2019-08-09T05:47:59.044Z] Read at 0x00c0011fc5a0 by goroutine 101:
[2019-08-09T05:47:59.044Z]   github.com/pingcap/tidb/distsql.(*selectResult).Close()
[2019-08-09T05:47:59.044Z]       /home/jenkins/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/distsql/select_result.go:261 +0x268
[2019-08-09T05:47:59.044Z]   github.com/pingcap/tidb/executor.closeAll()
[2019-08-09T05:47:59.044Z]       /home/jenkins/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/executor/distsql.go:119 +0x11b
[2019-08-09T05:47:59.044Z]   github.com/pingcap/tidb/executor.(*TableReaderExecutor).Close()
[2019-08-09T05:47:59.044Z]       /home/jenkins/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/executor/table_reader.go:251 +0x3fe
[2019-08-09T05:47:59.045Z]   github.com/pingcap/tidb/executor.(*baseExecutor).Close()
[2019-08-09T05:47:59.045Z]       /home/jenkins/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/executor/executor.go:110 +0xaf
[2019-08-09T05:47:59.045Z]   github.com/pingcap/tidb/executor.(*HashJoinExec).Close()
[2019-08-09T05:47:59.045Z]       /home/jenkins/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/executor/join.go:138 +0x433
[2019-08-09T05:47:59.045Z]   github.com/pingcap/tidb/executor.(*DeleteExec).Close()
[2019-08-09T05:47:59.045Z]       /home/jenkins/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/executor/delete.go:178 +0x76
[2019-08-09T05:47:59.045Z]   github.com/pingcap/tidb/executor.(*ExecStmt).handleNoDelayExecutor.func1()
[2019-08-09T05:47:59.045Z]       /home/jenkins/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/executor/adapter.go:420 +0x3e
[2019-08-09T05:47:59.045Z]   github.com/pingcap/tidb/executor.(*ExecStmt).handleNoDelayExecutor()
[2019-08-09T05:47:59.045Z]       /home/jenkins/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/executor/adapter.go:428 +0x3b1
[2019-08-09T05:47:59.045Z]   github.com/pingcap/tidb/executor.(*ExecStmt).handleNoDelay()
[2019-08-09T05:47:59.045Z]       /home/jenkins/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/executor/adapter.go:297 +0x1a3
[2019-08-09T05:47:59.045Z]   github.com/pingcap/tidb/executor.(*ExecStmt).Exec()
[2019-08-09T05:47:59.045Z]       /home/jenkins/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/executor/adapter.go:265 +0x33f
[2019-08-09T05:47:59.045Z]   github.com/pingcap/tidb/session.runStmt()
[2019-08-09T05:47:59.045Z]       /home/jenkins/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/session/tidb.go:225 +0x1a1
[2019-08-09T05:47:59.045Z]   github.com/pingcap/tidb/session.(*session).executeStatement()
[2019-08-09T05:47:59.045Z]       /home/jenkins/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/session/session.go:1000 +0x203
[2019-08-09T05:47:59.045Z]   github.com/pingcap/tidb/session.(*session).execute()
[2019-08-09T05:47:59.045Z]       /home/jenkins/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/session/session.go:1111 +0xbd7
[2019-08-09T05:47:59.045Z]   github.com/pingcap/tidb/session.(*session).Execute()
[2019-08-09T05:47:59.045Z]       /home/jenkins/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/session/session.go:1039 +0xd4
[2019-08-09T05:47:59.045Z]   github.com/pingcap/tidb/util/testkit.(*TestKit).Exec()
[2019-08-09T05:47:59.045Z]       /home/jenkins/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/util/testkit/testkit.go:143 +0x103
[2019-08-09T05:47:59.045Z]   github.com/pingcap/tidb/util/testkit.(*TestKit).MustExec()
[2019-08-09T05:47:59.045Z]       /home/jenkins/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/util/testkit/testkit.go:181 +0x91
[2019-08-09T05:47:59.045Z]   github.com/pingcap/tidb/sessionctx/binloginfo_test.(*testBinlogSuite).TestDeleteSchema()
[2019-08-09T05:47:59.045Z]       /home/jenkins/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/sessionctx/binloginfo/binloginfo_test.go:432 +0x1f1
[2019-08-09T05:47:59.045Z]   github.com/pingcap/tidb/sessionctx/binloginfo_test.(*testBinlogSuite).TestDeleteSchema()
[2019-08-09T05:47:59.045Z]       /home/jenkins/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/sessionctx/binloginfo/binloginfo_test.go:425 +0x121
[2019-08-09T05:47:59.045Z]   github.com/pingcap/tidb/sessionctx/binloginfo_test.(*testBinlogSuite).TestDeleteSchema()
[2019-08-09T05:47:59.045Z]       /home/jenkins/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/sessionctx/binloginfo/binloginfo_test.go:424 +0xed
[2019-08-09T05:47:59.045Z]   runtime.call32()
[2019-08-09T05:47:59.045Z]       /usr/local/go/src/runtime/asm_amd64.s:519 +0x3a
[2019-08-09T05:47:59.045Z]   reflect.Value.Call()
[2019-08-09T05:47:59.045Z]   reflect.Value.Call()
[2019-08-09T05:47:59.045Z]       /usr/local/go/src/reflect/value.go:308 +0xc0
[2019-08-09T05:47:59.045Z]   github.com/pingcap/check.(*suiteRunner).forkTest.func1()
[2019-08-09T05:47:59.045Z]       /home/jenkins/workspace/tidb_ghpr_unit_test/go/pkg/mod/github.com/pingcap/check@v0.0.0-20190102082844-67f458068fc8/check.go:836 +0x9fc
[2019-08-09T05:47:59.045Z]   github.com/pingcap/check.(*suiteRunner).forkCall.func1()
[2019-08-09T05:47:59.045Z]       /home/jenkins/workspace/tidb_ghpr_unit_test/go/pkg/mod/github.com/pingcap/check@v0.0.0-20190102082844-67f458068fc8/check.go:730 +0xb7
[2019-08-09T05:47:59.045Z] 
[2019-08-09T05:47:59.045Z] Previous write at 0x00c0011fc5a0 by goroutine 84:
[2019-08-09T05:47:59.045Z]   github.com/pingcap/tidb/distsql.(*selectResult).getSelectResp()
[2019-08-09T05:47:59.045Z]       /home/jenkins/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/distsql/select_result.go:176 +0xaa0
[2019-08-09T05:47:59.045Z]   github.com/pingcap/tidb/distsql.(*selectResult).Next()
[2019-08-09T05:47:59.045Z]       /home/jenkins/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/distsql/select_result.go:149 +0xf0
[2019-08-09T05:47:59.045Z]   github.com/pingcap/tidb/executor.(*tableResultHandler).nextChunk()
[2019-08-09T05:47:59.045Z]       /home/jenkins/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/executor/table_reader.go:229 +0xaa
[2019-08-09T05:47:59.045Z]   github.com/pingcap/tidb/executor.(*TableReaderExecutor).Next()
[2019-08-09T05:47:59.045Z]       /home/jenkins/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/executor/table_reader.go:149 +0x20c
[2019-08-09T05:47:59.045Z]   github.com/pingcap/tidb/executor.(*HashJoinExec).fetchInnerRows()
[2019-08-09T05:47:59.045Z]       /home/jenkins/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/executor/join.go:275 +0x750
[2019-08-09T05:47:59.045Z]   github.com/pingcap/tidb/executor.(*HashJoinExec).fetchInnerAndBuildHashTable.func1()
[2019-08-09T05:47:59.045Z]       /home/jenkins/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/executor/join.go:539 +0x65
[2019-08-09T05:47:59.045Z]   github.com/pingcap/tidb/util.WithRecovery()
[2019-08-09T05:47:59.045Z]       /home/jenkins/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/util/misc.go:81 +0x5a
[2019-08-09T05:47:59.045Z] 

```

2. What did you expect to see?

CI should pass

3. What did you see instead?

 DATA RACE

4. What version of TiDB are you using (`tidb-server -V` or run `select tidb_version();` on TiDB)?

Basically, this one 036e7e207deb4b58a98bdea2c3bf500a2188d3f2

**Comment by SunRunAway** [comment]
[log.txt](https://github.com/pingcap/tidb/files/3484967/log.txt)
